README.md
java-danmaku-engine
===================

Danmaku engine built on Java, OpenGL & OpenAL, currently on alpha

·Coding: HarZe - https://github.com/HarZe

·Artwork: sprites, sound effects and music extracted from several games of http://en.wikipedia.org/wiki/Touhou_Project
- Scalation of some Touhou's artwork by:
http://popfan95b.deviantart.com/ - http://popfan95b.deviantart.com/art/Touhou-Kokichi-Spritework-Bullets-435884754


JDE/changelog
Version 0.2.1 (October 22, 2014):
Added Launcher menu
Added fixed FPS (forces and limits FPS to a given target)
Added real FPS counter
Added "rotation" to <sprite>
Now <hud> is optional in <game>, default HUD is the new JDE-HUD
Update license to add CC BY-NC v4.0 to all the multimedia and XML files

Version 0.2 (October 19, 2014):
Pre-alpha officially finished, beginning alpha development
Added <sound> and <music> to play sounds in-game
Added some Touhou SFX and music to use in the example, including XML
Added <sound> to <bullet> for spawning sound effect
Added <sound> and <music> to <wave>, in order to change music and/or set spawning wave sound effect
Added <sound> to <enemy> for spawning sound effect
Added <music> to <game> to include background music
Added attribute "power" to <bullet>, to measure the damage dealt
Added attribute "respawn-time" to <player>, to specify the invulnerability time after respawn
Added attribute "lifes" to <player>, which counts the remaining lifes when starting
Added attribute "points" to <enemy>, the score points got when destroying it
Added attribute "layer" to <sprite-label> and <text-label> to manage overlap
Player's bullets now deal damage to enemies (and vanish) and destroy them eventually
Added simple score system
Lots and lots of bug fixes and minor changes

Version 0.1.11 (September 26, 2014):
Added <hud>, <sprite-label>, <text-label> and <pseudo-font> to create a custom HUD via XML
More XML bug fixes (mostly error checking missing)
Added a basic XML example of <hud>
Rename "test.xml" to "touhou_example.xml"
Added Javadoc for all new content

Version 0.1.10 (September 15, 2014):
Updated LICENSE to GPLv3
Lots of XML format bugs fixed
XML format changes in <animation>
Parser documented (with Javadoc)

Version 0.1.9 (September 10, 2014):
Fixed player not sliding on game board border
Added <wave> to <player> in order to attack with bullets (colliding with enemies not implemented yet)
Added attribute "cooldown" to <player> in order to regulate the "time rest" between attacking waves
Added Javadoc to Player

Version 0.1.8 (September 9, 2014):
Added custom resolution
Added JDE icon/logo

Version 0.1.7 (September 7, 2014):
Holydays finished (uh, forget that one)
Added Javadoc for almost all classes
Some minor code optimizations

Version 0.1.6 (August 14, 2014):
Added all Touhou bullet's sprites and almost all hitboxes
Fixed "rotation" not being applied in last <direction> of a <movement>
Fixed "look-at-direction" value not working in case of "no"

Version 0.1.5 (August 11, 2014):
Added "homing-offset" to <direction>, and "homing-offset-start" and "homing-offset-end" to <direction-modifier> (default: 0)
Added "homing-to-player" to <direction> (default: no)
Changed "repeat" to "loop-from" in <animation>
Added two optional animations to <player> for moving left and right
Added Touhou's Reimu player animations

Version 0.1.4 (August 10, 2014):
Added collision with enemies
Modified <player> to get only <vertex> as spawn point, and not a <movement>
Added "hit-size" to <player> to set the radius of player hitbox, to "inflate" all hit-bodies in the game
Fixed some bullets not colliding because multithreading

Version 0.1.3 (August 10, 2014):
Fixed a mayor bug in sprite loading, now FPS are 1000% higher
Fixed angle corretion for hit-polygon rotation
Added basic multithread to draw and collide more efficiently

Version 0.1.2 (August 10, 2014):
Fixed "inside-checking" in hit-polygons
Fixed a critical bug about cyrus-beck algorith

Version 0.1.1 (August 10, 2014):
Fixed some internal errors

Version 0.1 (Agusut 9, 2014):
Added collision system for circles and convex polygons: <hit-circle> and <hit-polygon>
Added <hit-body> to contain multiple <hit-circle> or <hit-polygon> into an entity
Added parameter "spin" to <movement> to get local rotation angle, default: 90 (general correction) 
Added parameter "look-at-direction" to <movement> to specify if sprite rotates at where is moving, default: yes (for bullets), enemies use to have "no"
Added <animation> which contains a set of <sprites> with a duration, animations can loop, now the entities have <animation> instead of <sprite>
Added <player> with control with arrows and focus with left SHIFT

Version 0.06.1 (August 5, 2014):
Fixed <direction-modifier> not having motionStart and motionEnd
Added "exponent" attribute for each basic atribute of <direction> in the <direction-modifier>

Version 0.06 (August 5, 2014):
Changed <wave> attribute "repeat" to be a number of repetitions, <0 for infinite, default: 0
Added inheritance disable option to the <direction>
Added <wave> attribute "exponent" that allows to curve the time used for bullet spawning, default: 1

Version 0.05.1 (July 28, 2014):
Fixed wave time calculation not working properly
Modified XML sample

Version 0.05 (July 27, 2014):
Renamed XML attributes with multiple words for a better reading
Added <wave> absolute spawn position
Added <wave> attribute: waves (to repeat all the waves counting)
Added new <direction> attributes:
- motion: rotation acceleration
- random offsets for all the attributes (default 0)

Version 0.04.1 (July 25, 2014):
Added XML import duplication protection
Fixed attributes with "start" and "end" values not using the "end" value in last iteration
Rotation is now measured in degrees/second

Version 0.04 (July 21, 2014):
Removed <horde> and <wave> takes its place, now <wave> can do its job
Now <wave> can use timestart and timeend on other <wave>
Combined XML samples, still a lot to test

Version 0.03.7 (July 20, 2014):
Added spawnpoint combination from <wave> to the insiders <wave>
New XML sample with all new features

Version 0.03.6 (July 20, 2014):
Now <wave> can inherit any time

Version 0.03.5 (July 20, 2014):
Added new type of <wave> that contains more <wave>, still not fully working, it only inherits once

Version 0.03.4 (July 20, 2014):

JDE/src/com/jde/view/Launcher.java
package com.jde.view;

import java.awt.EventQueue;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;

import javax.swing.ButtonGroup;
import javax.swing.DefaultComboBoxModel;
import javax.swing.JButton;
import javax.swing.JComboBox;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JRadioButton;
import javax.swing.JTextField;

import org.lwjgl.LWJGLException;

import com.jde.model.utils.Parser;

public class Launcher {

	/**
	 * Launch the application.
	 */
	public static void launch() {
		EventQueue.invokeLater(new Runnable() {
			public void run() {
				try {
					Launcher window = new Launcher();
					window.frmJavaDanmakuEngine.setVisible(true);
				} catch (Exception e) {
					e.printStackTrace();
				}
			}
		});
	}
	private JFrame frmJavaDanmakuEngine;
	private JTextField textFieldWidthRes;
	private JTextField textFieldHeightRes;

	private final ButtonGroup buttonGroup = new ButtonGroup();

	/**
	 * Create the application.
	 */
	public Launcher() {
		initialize();
	}

	/**
	 * Initialize the contents of the frame.
	 */
	private void initialize() {
		
		// Detect XML game files
		File folder = new File("res/games/");
		ArrayList<String> listOfFiles = new ArrayList<String>();

		for (File f : folder.listFiles())
		    if (f.isFile() && f.getName().endsWith(".xml"))
		    	listOfFiles.add(f.getName());
		
		String[] arrayOfFiles = new String[listOfFiles.size()];
		int i = 0;
		for (String s : listOfFiles)
			arrayOfFiles[i++] = s;
		
			
		// Swing
		frmJavaDanmakuEngine = new JFrame();
		frmJavaDanmakuEngine.setResizable(false);
		frmJavaDanmakuEngine.setTitle("Java Danmaku Engine" + Game.VERSION + " - Launcher");
		frmJavaDanmakuEngine.setBounds(100, 100, 480, 240);
		frmJavaDanmakuEngine.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
		frmJavaDanmakuEngine.getContentPane().setLayout(null);
		
		final JComboBox<String> comboBoxResolution = new JComboBox<String>();
		comboBoxResolution.setToolTipText("Those are the most used 4:3 default resolutions");
		comboBoxResolution.setModel(new DefaultComboBoxModel<String>(new String[] {"640x480", "960x720", "1280x960"}));
		comboBoxResolution.setSelectedIndex(1);
		comboBoxResolution.setBounds(191, 7, 271, 24);
		frmJavaDanmakuEngine.getContentPane().add(comboBoxResolution);
		
		final JRadioButton rdbtnDefaultRes = new JRadioButton("Default resolution:");
		buttonGroup.add(rdbtnDefaultRes);
		rdbtnDefaultRes.setSelected(true);
		rdbtnDefaultRes.setBounds(8, 8, 175, 23);
		frmJavaDanmakuEngine.getContentPane().add(rdbtnDefaultRes);
		
		JRadioButton rdbtnCustomRes = new JRadioButton("Custom resolution:");
		buttonGroup.add(rdbtnCustomRes);
		rdbtnCustomRes.setBounds(8, 35, 175, 23);
		frmJavaDanmakuEngine.getContentPane().add(rdbtnCustomRes);
		
		textFieldWidthRes = new JTextField();
		textFieldWidthRes.setBounds(191, 37, 114, 19);
		frmJavaDanmakuEngine.getContentPane().add(textFieldWidthRes);
		textFieldWidthRes.setColumns(10);
		
		textFieldHeightRes = new JTextField();
		textFieldHeightRes.setBounds(348, 37, 114, 19);
		frmJavaDanmakuEngine.getContentPane().add(textFieldHeightRes);
		textFieldHeightRes.setColumns(10);
		
		JLabel lblX = new JLabel("x");
		lblX.setBounds(323, 37, 27, 19);
		frmJavaDanmakuEngine.getContentPane().add(lblX);
		
		JLabel lblFixedFps = new JLabel("Fixed FPS:");
		lblFixedFps.setBounds(8, 66, 90, 15);
		frmJavaDanmakuEngine.getContentPane().add(lblFixedFps);
		
		final JComboBox<String> comboBoxFixedFps = new JComboBox<String>();
		comboBoxFixedFps.setToolTipText("<html>This option forces the computer to play at specific frame rate (60 for example),<br>and the physics engine has 3 times more frames (180 in the example) to improve accuracy.<br>Also, if your computer cannot keep up the target FPS, the game will slow down to avoid skipping frames.<br><br>When \"Unlimited\" is selected, the computer will process as much frames as it can.<br>The game will not slow down if need. This mode is not recommended,<br>because fps drops can be very unestable for the engine.<br>But it is useful to check how much fixed FPS your computer is able to handle.</html>");
		comboBoxFixedFps.setModel(new DefaultComboBoxModel<String>(new String[] {"30 FPS [90 engine FPS] (not recommended)", "60 FPS [180 engine FPS] (recommended for average PCs)", "120 FPS [360 engine FPS] (recommended for high end PCs)", "Unlimited (not recommended, unstable)"}));
		comboBoxFixedFps.setSelectedIndex(1);
		comboBoxFixedFps.setBounds(8, 88, 454, 24);
		frmJavaDanmakuEngine.getContentPane().add(comboBoxFixedFps);
		
		JLabel lblAvailableGames = new JLabel("Available games:");
		lblAvailableGames.setBounds(12, 124, 450, 15);
		frmJavaDanmakuEngine.getContentPane().add(lblAvailableGames);
		
		final JComboBox<String> comboBoxAvailableGames = new JComboBox<String>();
		comboBoxAvailableGames.setBounds(8, 146, 454, 24);
		comboBoxAvailableGames.setModel(new DefaultComboBoxModel<String>(arrayOfFiles));
		frmJavaDanmakuEngine.getContentPane().add(comboBoxAvailableGames);
		
		JButton btnPlay = new JButton("Play");
		btnPlay.addMouseListener(new MouseAdapter() {
			@Override
			public void mouseClicked(MouseEvent e) {
				
				// Check resolution
				int w = 640, h = 480;
				
				if (rdbtnDefaultRes.isSelected()) {
					switch (comboBoxResolution.getSelectedIndex()) {
						
					case 1:
						w = 960;
						h = 720;
						break;
						
					case 2:
						w = 1280;
						h = 960;
						break;
					}
				}
				else {
					w = Integer.parseInt(textFieldWidthRes.getText());
					h = Integer.parseInt(textFieldHeightRes.getText());
				}
				
				// Check fixed FPS
				int fps = 60;
				
				switch (comboBoxFixedFps.getSelectedIndex()) {
				case 0:
					fps = 30;
					break;
				case 2:
					fps = 120;
					break;
				case 3:
					fps = 0;
					break;
				}
				
				// Start playing
				Game newGame;
				
				try {
					Parser parser = new Parser();
					newGame = parser.parseXML("res/games/" + comboBoxAvailableGames.getSelectedItem());
					new Window(newGame, w, h, fps);
					
				} catch (LWJGLException ex) {
					ex.printStackTrace();
				} catch (NumberFormatException ex) {
					ex.printStackTrace();
				} catch (IOException ex) {
					ex.printStackTrace();
				}
				
			}
		});
		btnPlay.setBounds(8, 182, 454, 25);
		frmJavaDanmakuEngine.getContentPane().add(btnPlay);
	}
}
